cpp/include/cudf/column/column_device_view.cuh

```diff
@@ -906,6 +976,119 @@ struct value_accessor {
   __device__ T operator()(cudf::size_type i) const { return col.element<T>(i); }
 };
 
+template <typename T, enum optional_mode>
+struct optional_accessor;
+
+/**
+ * @brief RUNTIME optional accessor of a maybe-nullable column
+ *
+ * The `RUNTIME` specialization of optional_accessor returns a
+ * thrust::optional of the scalar value at `id`.
+ *
+ * When `with_nulls=true` the return value will be a `thrust::optional{column[i]}`
+ * when column[i] is not null, and `thrust::optional{}` when column[i] is null.
+ *
+ * When `with_nulls=false` the return value will always be `thrust::optional{column[i]}`
+ *
+ * @throws cudf::logic_error if the column is not nullable, and `with_nulls=true`
+ * @throws cudf::logic_error if column datatype and template T type mismatch.
+ *
+ * @tparam T The type of elements in the column
+ */
+template <typename T>
+struct optional_accessor<T, optional_mode::RUNTIME> {
+  column_device_view const col;  ///< column view of column in device
+  bool has_nulls;
+
+  /**
+   * @brief constructor
+   * @param[in] _col column device view of cudf column
+   */
+  optional_accessor(column_device_view const& _col, bool with_nulls)
+    : col{_col}, has_nulls{with_nulls}
+  {
+    CUDF_EXPECTS(type_id_matches_device_storage_type<T>(col.type().id()), "the data type mismatch");
+    if (with_nulls) { CUDF_EXPECTS(_col.nullable(), "Unexpected non-nullable column."); }
+  }
+
+  CUDA_DEVICE_CALLABLE
+  thrust::optional<T> operator()(cudf::size_type i) const
+  {
+    return (has_nulls and !col.is_valid_nocheck(i)) ? thrust::optional<T>{thrust::nullopt}
+                                                    : thrust::optional<T>{col.element<T>(i)};
+  }
+};
+
+/**
+ * @brief NO_NULLS optional accessor of a column
+ *
+ * The `NO_NULLS` specialization of optional_accessor returns a
+ * thrust::optional of the scalar value at `id`.
+ *
+ * Since the user has attested that the column has no null values,
+ * no checks will occur and `thrust::optional{column[i]}` will be
+ * return for each `i`.
+ *
+ * @throws cudf::logic_error if column datatype and template T type mismatch.
+ *
+ * @tparam T The type of elements in the column
+ */
+template <typename T>
+struct optional_accessor<T, optional_mode::NO_NULLS> {
+  column_device_view const col;  ///< column view of column in device
+
+  /**
+   * @brief constructor
+   * @param[in] _col column device view of cudf column
+   */
+  optional_accessor(column_device_view const& _col, bool) : col{_col}
```

Comment:
I think the `NO_NULLS/HAS_NULLS` specializations shouldn't have the `bool` in their constructor parameter list. Otherwise you're forced to pass a `bool` in the ctor even when the template information already tells you bools do/do not exist.